web/popminer/dispatch.go

```diff
@@ -0,0 +1,438 @@
+	Required []DispatchArgs
+}
+
+func dispatch(this js.Value, args []js.Value) any {
+	defer func() {
+		if r := recover(); r != nil {
+			log.Criticalf("recovered panic: %v", r)
+			log.Criticalf(string(debug.Stack()))
+		}
+	}()
+
+	log.Tracef("dispatch")
+	defer log.Tracef("dispatch exit")
+
+	// Create JS Promise handler
+	handler := js.FuncOf(func(_ js.Value, handlerArgs []js.Value) any {
+		resolve := handlerArgs[0]
+		reject := handlerArgs[1]
+
+		// Run dispatched call asynchronously
+		go func() {
+			// This function must always complete a promise.
+			defer func() {
+				if r := recover(); r != nil {
+					log.Criticalf("recovered panic: %v", r)
+					log.Criticalf(string(debug.Stack()))
+					reject.Invoke(jsError(fmt.Errorf("recovered panic: %v", r)))
+				}
+			}()
+
+			// parse args
+			d, err := parseArgs(args)
+			if err != nil {
+				reject.Invoke(jsError(err))
+				return
+			}
+
+			// dispatch sanitized call
+			rv, err := d.Handler(this, args)
+			if err != nil {
+				reject.Invoke(jsError(err))
+				return
+			}
+			resolve.Invoke(rv)
+		}()
+
+		// The handler of a Promise doesn't return any value
+		return nil
+	})
+	defer handler.Release()
+
+	// Create and return the Promise object
+	return promiseConstructor.New(handler)
+}
+
+func parseArgs(args []js.Value) (*Dispatch, error) {
+	// Verify we received data readable command
+	if len(args) != 1 {
+		return nil, fmt.Errorf("expected 1 argument, got %v", len(args))
+	}
+	data := args[0]
+	if data.Type() != js.TypeObject {
+		return nil, fmt.Errorf("expected an object, got: %v", data.Type())
+	}
+
+	m := data.Get("method")
+	if m.Type() != js.TypeString {
+		return nil, fmt.Errorf("expected method to be a string, got: %v", m.Type())
+	}
+	d, ok := Dispatcher[m.String()]
+	if !ok {
+		log.Warningf("method not found: %v", m.String())
+		return nil, fmt.Errorf("method not found: %v", m.String())
+	}
+
+	// Verify required args
+	for k := range d.Required {
+		name := d.Required[k].Name
+		typ := d.Required[k].Type
+		arg := data.Get(name)
+		if arg.Type() != typ {
+			return nil, fmt.Errorf("%v: invalid type %v, want %v",
+				name, arg.Type(), typ)
+		}
+	}
+
+	return d, nil
+}
+
+func wasmVersion(_ js.Value, _ []js.Value) (js.Value, error) {
+	return Object{
+		"version":   version,
+		"gitCommit": gitCommit,
+	}.Value(), nil
+}
+
+func generateKey(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("generatekey")
+	defer log.Tracef("generatekey exit")
+
+	var (
+		btcChainParams *btcchaincfg.Params
+		netNormalized  string
+	)
+	net := args[0].Get("network").String()
+	switch net {
+	case "devnet", "testnet3", "testnet":
+		btcChainParams = &btcchaincfg.TestNet3Params
+		netNormalized = "testnet3"
+	case "mainnet":
+		btcChainParams = &btcchaincfg.MainNetParams
+		netNormalized = "mainnet"
+	default:
+		return js.Null(), fmt.Errorf("invalid network: %v", net)
+	}
+
+	privKey, err := dcrsecpk256k1.GeneratePrivateKey()
```

Comment:
Not sure we want to pull in all this stuff, including some btcuti/ethereum stuff. For now this is fine but we need to see if we can pull it out and save some valuable network bandwidth.

Reply:
Added a comment.